crates/oxc_minifier/src/peephole/minimize_statements.rs

```diff
@@ -443,6 +446,254 @@ impl<'a> PeepholeOptimizations {
         }
     }
 
+    /// Determines whether an array destruction assignment can be simplified based on the provided variable declaration.
+    /// - `let [x, y] = [1, 2];` -> true
+    /// - `let [x, y] = [...arr];` -> false
+    fn can_simplify_array_to_array_destruction_assignment(
+        decl: &VariableDeclarator<'a>,
+        ctx: &Ctx<'a, '_>,
+    ) -> bool {
+        let BindingPattern::ArrayPattern(id_kind) = &decl.id else { return false };
+        // if left side of assignment is empty do not process it
+        if id_kind.is_empty() {
+            return false;
+        }
+
+        let Some(Expression::ArrayExpression(init_expr)) = &decl.init else { return false };
+
+        let first = init_expr.elements.first();
+
+        // check if the first init is not spread when rest is present without elements
+        // [] = [...rest] | [a, ...rest] = [...rest]
+        if first.is_some_and(ArrayExpressionElement::is_spread)
+            && id_kind.rest.is_none()
+            && !id_kind.elements.is_empty()
+        {
+            return false;
+        }
+
+        // check for `[a = b] = [c]`, this could be potentially optimized if we check if `c` is undefined
+        if init_expr.elements.len() == 1
+            && id_kind
+                .elements
+                .first()
+                .is_some_and(|e| e.as_ref().is_none_or(BindingPattern::is_assignment_pattern))
+            && first.is_none_or(|expr| expr.value_type(ctx).is_undetermined())
+        {
+            return false;
+        }
+
+        true
+    }
+
+    fn simplify_array_destruction_assignment(
+        decl: &mut VariableDeclarator<'a>,
+        result: &mut Vec<'a, VariableDeclarator<'a>>,
+        ctx: &mut Ctx<'a, '_>,
+    ) -> bool {
+        let BindingPattern::ArrayPattern(id_pattern) = &mut decl.id else {
+            return false;
+        };
+        let Some(Expression::ArrayExpression(init_expr)) = &mut decl.init else {
+            return false;
+        };
+
+        // limit iteration of id (left) to last spread of init (right)
+        // [a, b, c] = [b, ...spread]
+        let index = if let Some(spread_index) =
+            init_expr.elements.iter().position(ArrayExpressionElement::is_spread)
+        {
+            min(id_pattern.elements.len(), spread_index)
+        } else {
+            id_pattern.elements.len()
+        };
+
+        if decl.kind.is_var() && index > 1 {
+            for init_val in &mut init_expr.elements {
+                if init_val.value_type(ctx).is_undetermined() {
```

Comment:
What is the purpose of this `if`? Is it to avoid unnecessary temporary vars? In that case, I think we should use `.is_literal_value(false, ctx)` instead. For example, `.value_type(ctx)` returns `ValueType::boolean` for `!a`.

Reply:
correct, there is no need to create variables for literals